std/traits.d

```diff
@@ -8087,3 +8089,97 @@ enum isCopyable(S) = is(typeof(
     static assert(isCopyable!int);
     static assert(isCopyable!(int[]));
 }
+
+/**
+ * Determines the type of the argument.
+ *
+ * Returns:
+ *   T[0] if it's a type, typeof(T[0]) if it's a value, void if it's a template.
+ *
+ */
+template TypeOf(T...)
+    if (T.length == 1 && (is(T[0]) || is(typeof(T[0]))))
+{
+    static if (is(T[0]))
+        alias TypeOf = T[0];
+    else static if (is(typeof(T[0])))
+        alias TypeOf = typeof(T[0]);
+}
+
+///
+@safe unittest
+{
+    assert(is(TypeOf!3 == int));
+    assert(is(TypeOf!int == int));
+    assert(is(TypeOf!([]) == void[]));
+    assert(!is(TypeOf!(mixin(__MODULE__))));
+    assert(is(TypeOf!TypeOf));
+}
+
+/**
+ * Get a list of the overloads for the given callable.
+ */
+template getOverloads(func...)
+    if (func.length == 1 && isCallable!func)
+{
+    alias p = AliasSeq!(__traits(parent, func))[0];
+    static if (is(p) || p.stringof[0 .. 7] == "module ")
+        alias getOverloads = AliasSeq!(__traits(getOverloads, __traits(parent, func), __traits(identifier, func)));
+    else
+        alias getOverloads = func;
+}
+
+version(unittest)
+{
+    int getOverloadsTest(int n = 3);
+    int getOverloadsTest(int, int);
+    int getOverloadsTest2(int, int);
+}
+
+///
+@safe unittest
+{
+    assert(getOverloads!getOverloadsTest.length == 2);
+    assert(is(typeof(&getOverloadsTest) == typeof(&getOverloads!getOverloadsTest[0])));
+    assert(!is(typeof(&getOverloadsTest) == typeof(&getOverloads!getOverloadsTest[1])));
+    alias fn = getOverloadsTest;
+    assert(is(typeof(&fn) == typeof(&getOverloads!getOverloadsTest[0])));
+    assert(!is(typeof(&fn) == typeof(&getOverloads!getOverloadsTest[1])));
+    assert(getOverloads!getOverloadsTest2.length == 1);
+    assert(getOverloads!(() => 3).length == 1);
+    assert(getOverloads!(() => 3).length == 1);
+}
+
+/**
+ * Gets the overload for `func` that can be called with `Args` as arguments, if any.
+ */
+template overloadFor(alias func, Args...)
+    if (isCallable!func)
+{
+    alias ArgTypes = staticMap!(TypeOf, Args);
+    template match(fns...)
+    {
+        static if (fns.length == 0)
+            static assert(false, "No overload found for parameter types "~Args.stringof);
```

Comment:
`static assert`-ing here seems incorrect on first glance. Why not return an empty `AliasSeq`?